Create builder function that creates NFA from regex

jurajsic opened the pull request:
Also it seems some profiling tests assumed that the created automaton is not reduced but it was, I changed it so it is actually not reduced. Maybe I should change it back?

jurajsic committed:
create_nfa from regex returns nfa instead of taking pointer as argument

jurajsic committed:
put back version with the argument for python bindings

jurajsic committed:
add builder function to create nfa from regex

Adda0 reviewed (commented):
I think all the uses of `parser::create_nfa()` in Mata and the user code should be replaced by `nfa::builder::create_from_regex()`

`nfa::builder::create_from_regex()` is more descriptive, much more visible (builder vs. parser), and most importantly, gives a reasonable interface for creating an NFA from a regex. `parser::create_nfa()` does not tell one anything, is hidden within the library, and I would consider it an implementation detail of parsing regexes, ideally to be hidden from the public library interface and easily replaceable with anything else.

jurajsic commented:
You mean also in tests (except the ones testing `create_nfa`)? I replaced some, then I realised I am not sure if I want to replace all.

We should also decide on the default encoding, whether it should be `Latin1` (right now) or `UTF8`.

Adda0 commented:
> You mean also in tests (except the ones testing `create_nfa`)? I replaced some, then I realised I am not sure if I want to replace all.

Yeah. Except for the tests for the actual implementation, I would replace all other uses.

> We should also decide on the default encoding, whether it should be Latin1 (right now) or UTF8.

We switched to the Latin1 encoding, as Latin1 uses direct encoding on 8 bits. See the PR https://github.com/VeriFIT/mata/pull/461 and its tests. Do you think UTF-8 would be a better default?

jurajsic commented:
> > You mean also in tests (except the ones testing `create_nfa`)? I replaced some, then I realised I am not sure if I want to replace all.
> 
> Yeah. Except for the tests for the actual implementation, I would replace all other uses.

Ok I will replace it.

> > We should also decide on the default encoding, whether it should be Latin1 (right now) or UTF8.
> 
> We switched to the Latin1 encoding, as Latin1 uses direct encoding on 8 bits. See the PR #461 and its tests. Do you think UTF-8 would be a better default?

Ah, yeah, I forgot. I will add a comment for this.

jurajsic commented:
I think I should also add some comment to `nfa::builder::create_from_regex()` that the regexes actually do not support everything that RE2 does, with some mention of issues #437, #457 and #464 (not sure if all these issues are still issues).

Adda0 commented:
Except for https://github.com/VeriFIT/mata/issues/464 (discusses a possible extension to the regex parser which could modify how the same regex can be interpreted; it is actually pretty useful, just not implemented as of now), the remaining issues should be fixed. I just have not verified that is indeed fixed so that we can close the issues. When someone finds some time, they need to go through all the examples in the other two issues and verify that everything works now.

jurajsic committed:
add better comments explaining limitations

jurajsic committed:
reference to functions in comments

jurajsic committed:
use create_from_regex instead of create_nfa in tests

jurajsic commented:
I made the changes, I kept `parser::create_nfa()` in `nfa-profiling.cc` tests, as it seems they are needed there. Can you check this file if my changes are ok? Because I changed the arguments there too.

jurajsic committed:
add test cases for issues #457 and #437

jurajsic commented:
Also, I added test cases for #437 and #457, seems like the issues are not fixed yet (so they are commented out for now).

Adda0 commented on tests/nft/nft.cc:
```diff
@@ -3487,19 +3483,6 @@ TEST_CASE("mata::nft::get_useful_states_tarjan") {
 		CHECK(bv == ref);
 	}
 
-    SECTION("from regex (a+b*a*)") {
-        Nft aut;
-        mata::parser::create_nfa(&aut, "(a+b*a*)", false, EPSILON, false);
-
-        mata::BoolVector bv = aut.get_useful_states();
-        mata::BoolVector ref({ 1, 0, 1, 0, 1, 0, 1, 0, 0});
-        CHECK(bv == ref);
-
-        aut = reduce(aut.trim());
-        bv = aut.get_useful_states();
-        CHECK(bv == mata::BoolVector({ 1, 1, 1, 1}));
-    }
-
```
Why was this test removed entirely? While I agree that this is a pretty bad test, we could at least rewrite it into a test with more useful and sensible checks.

(This comment should be for the NFA tests/nfa/nfa.cc equivalent for the test.)

jurajsic replied:
It must use `create_nfa` as it does not want to reduce the NFA created from the regex, so I couldn't replace with `create_from_regex`. However, it also expects a specific structure of the created NFA from a regex, which is weird, it assumes that it works specific to RE2 parser.

Adda0 replied:
Yeah, the checked conditions are silly and check implementation details instead of any reasonable qualities of the regex. OK, let us remove the test altogether. I do not think there is anything special not already checked by the other tests in this particular regex.

Adda0 reviewed (approved):
Great changes.
